src/chainstate/stacks/boot/cost-voting.clar

```diff
@@ -0,0 +1,208 @@
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err 1))))
+    )
+    (if (and
+            (< burn-block-height expiration-block-height)
+            (is-none (map-get? proposal-confirmed-id { proposal-id: proposal-id }))
+        )
+        (begin
+            (unwrap! (stx-transfer? amount tx-sender (as-contract tx-sender)) (err 1))
+            (unwrap! (ft-mint? cost-vote-token amount tx-sender) (err 1))
+            (map-set proposal-votes { proposal-id: proposal-id } { votes: (+ amount cur-votes) })
+            (map-set principal-proposal-votes { address: tx-sender, proposal-id: proposal-id}
+                                            { votes: (+ amount cur-principal-votes)})
+            (ok true))
+    (err 1)))
+)
+
+;; Withdraw votes
+(define-public (withdraw-votes (proposal-id uint) (amount uint))
+    (let (
+        (cur-votes (default-to u0 (get votes (map-get? proposal-votes { proposal-id: proposal-id }))))
+        (cur-principal-votes (default-to u0 (get votes (map-get? principal-proposal-votes {
+            address: tx-sender,
+            proposal-id: proposal-id }))))
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err 1))))
+        (sender tx-sender)
+    )
+    (if (and
+            (>= cur-principal-votes amount)
+            (< burn-block-height expiration-block-height)
+            ;;(is-none (map-get? confirmed-proposals { proposal-id: proposal-id }))
+        )
+        (begin
+            (unwrap! (as-contract (stx-transfer? amount tx-sender sender)) (err 1))
+            (unwrap! (as-contract (ft-transfer? cost-vote-token amount sender tx-sender)) (err 1))
+            (map-set proposal-votes { proposal-id: proposal-id } { votes: (- cur-votes amount) })
+            (map-set principal-proposal-votes { address: tx-sender, proposal-id: proposal-id }
+                                              { votes: (- cur-principal-votes amount) })
+            (ok true))
+        (err 1)))
+)
+
+;; Withdraw STX after vote is over
+(define-public (withdraw-after-votes (proposal-id uint))
+    (let (
+        (cur-principal-votes (default-to u0 (get votes (map-get? principal-proposal-votes {
+            address: tx-sender,
+            proposal-id: proposal-id }))))
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err 1))))
+        (sender tx-sender)
+    )
+    (if (> burn-block-height expiration-block-height)
+        (begin
+            (unwrap! (as-contract (stx-transfer? cur-principal-votes tx-sender sender)) (err 1))
+            (unwrap! (as-contract (ft-transfer? cost-vote-token cur-principal-votes sender tx-sender)) (err 1))
+            (ok true))
+        (err 1)))
+)
+
+;; Miner veto
+(define-public (veto (proposal-id uint))
```

Comment:
I think you need to add a data-map like `(define-data-map excercised-vetos ((proposol-id uint) (veto-height uint)) ((vetoed bool)))` which you can use to ensure that at any given stacks block height, at most one veto has been issued per proposal.